File: src/Orleans.Dataflow.FSharp/Fork.fs
namespace Orleans.Dataflow.FSharp

open System
open Orleans.Dataflow.Authoring

// Orleans.Dataflow itself is deliberately not opened: see the note in Source.fs.

/// <summary>Rejoins the two derived streams of a diamond.</summary>
/// <remarks>
/// <para>
/// A fork has two open ends and no way to close a graph, so a program that builds one has to rejoin it —
/// which is what makes these the only two functions the type needs. Both are total and positional: one
/// element into the fork produces one element out, built from that element's two derivations and never from
/// two different elements'. That is what separates this join from
/// <see cref="M:Orleans.Dataflow.FSharp.Source.zip``2"/> over two unrelated sources, which is positional and
/// nothing more.
/// </para>
/// <para>
/// <see cref="M:Orleans.Dataflow.FSharp.Source.forkMerge``2"/> is the other rejoin, for when the two paths
/// produce one element type and the answer wanted is whichever arrives first. It lives on the source because
/// it never produces a fork value at all.
/// </para>
/// </remarks>
[<RequireQualifiedAccess>]
module Fork =

    /// <summary>Rejoins the two derived streams through a function of both.</summary>
    /// <param name="combine">The function building one element from the two derivations of one input element.</param>
    /// <param name="fork">The diamond to close, which is unchanged.</param>
    /// <returns>A source of one element per element the fork was fed.</returns>
    /// <remarks>
    /// The function never enters the document — a local graph states which stage stands where and never how
    /// an element was built — so a graph rejoined this way is <c>nondeployable</c> exactly as every other
    /// local graph holding a lambda is.
    /// </remarks>
    let zipWith (combine: 'T1 -> 'T2 -> 'Out) (fork: Fork<'T1, 'T2>) : Source<'Out> =
        Source<'Out>(
            fork.State.Combine(
                LocalStageDescriptor.Zip(LocalRowCombiner.Of(Func<'T1, 'T2, 'Out> combine)),
                LocalJunctionGuard.FanInPorts LocalVocabulary.MinFanIn))

    /// <summary>Rejoins the two derived streams into a stream of pairs.</summary>
    /// <param name="fork">The diamond to close, which is unchanged.</param>
    /// <returns>A source of one pair per element the fork was fed.</returns>
    /// <remarks>
    /// The pair names its halves by the order the fork was written in: the first member is the left flow's
    /// element and the second is the right flow's. It is a struct tuple because that is the very
    /// <see cref="T:System.ValueTuple`2"/> the C# facade's own rejoin produces, so one graph authored in
    /// either language carries one element type as well as one document.
    /// </remarks>
    let zip (fork: Fork<'T1, 'T2>) : Source<struct ('T1 * 'T2)> =
        zipWith (fun first second -> struct (first, second)) fork

File: src/Orleans.Dataflow/Fork.cs
using System.Globalization;
using Orleans.Dataflow.Authoring;

namespace Orleans.Dataflow;

/// <summary>
/// A diamond in flight: one stream broadcast through two flows, waiting for the call that rejoins them.
/// </summary>
/// <typeparam name="T1">The element type the left flow produces.</typeparam>
/// <typeparam name="T2">The element type the right flow produces.</typeparam>
/// <remarks>
/// <para>
/// A fork is the one authoring value with two open ends, and that is the whole reason it exists. Everything
/// else in this surface has one: a <see cref="Source{T}"/> is one stream, a <see cref="Branch{TIn}"/> is one
/// leg that ends in a sink, and a junction call takes a graph from one shape to another without ever leaving
/// two ends dangling. Re-convergence — the same elements going two ways and meeting again — cannot be
/// written as a tree, so it gets a carrier instead of a builder.
/// </para>
/// <para>
/// The rejoin is positional and total: <see cref="Zip()"/> pairs the two derived streams element by element,
/// which is legal without a buffer between them exactly because both sides descend from one broadcast and
/// therefore advance together. <see cref="Source{T}.ForkMerge"/> is the other rejoin, for when the two paths
/// produce the same element type and the answer wanted is whichever arrives first.
/// </para>
/// <para>
/// A fork is a value like every other: rejoining it twice builds two graphs, and neither disturbs the other.
/// </para>
/// </remarks>
public sealed class Fork<T1, T2>
{
    /// <summary>Initializes a new instance of the <see cref="Fork{T1, T2}"/> class.</summary>
    /// <param name="shape">The shape of the source, the broadcast, and the two flows, with two open ends.</param>
    internal Fork(LocalGraphShape shape) => Shape = shape;

    /// <summary>Gets the partial graph this fork carries.</summary>
    /// <value>The occurrences and wiring so far, with the two derived streams still open.</value>
    internal LocalGraphShape Shape { get; }

    /// <summary>Rejoins the two derived streams into a stream of pairs.</summary>
    /// <returns>A source of one pair per element the fork was fed.</returns>
    /// <remarks>
    /// The tuple names the halves in the order the fork was written, so <c>First</c> is the left flow's
    /// element and <c>Second</c> is the right flow's. One element in produces one pair out, and the pair is
    /// built from the two derivations of that very element, which is what makes this join deterministic
    /// where a zip of two unrelated sources is only positional.
    /// </remarks>
    public Source<(T1 First, T2 Second)> Zip() =>
        Zip(static (first, second) => (first, second));

    /// <summary>Rejoins the two derived streams through a function of both.</summary>
    /// <typeparam name="TOut">The element type the function produces.</typeparam>
    /// <param name="combine">The function building one element from the two derivations of one input.</param>
    /// <returns>A source of one element per element the fork was fed.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="combine"/> is <see langword="null"/>.</exception>
    /// <remarks>
    /// The same join as <see cref="Zip()"/> with the pair built by the author instead of by the tuple, which
    /// is what keeps a row-building fork from allocating a tuple only to take it apart again.
    /// </remarks>
    public Source<TOut> Zip<TOut>(Func<T1, T2, TOut> combine)
    {
        ArgumentNullException.ThrowIfNull(combine);

        return new Source<TOut>(Shape.Combine(
            LocalStageDescriptor.Zip(LocalRowCombiner.Of(combine)),
            LocalJunctionGuard.FanInPorts(LocalVocabulary.MinFanIn)));
    }

    /// <summary>Returns a one-line diagnostic summary of this fork.</summary>
    /// <returns>Text of the form <c>fork (4 stages)</c>, singular for one (<c>fork (1 stage)</c>).</returns>
    /// <remarks>The count is formatted with the invariant culture, and the method never throws.</remarks>
    public override string ToString() =>
        string.Create(
            CultureInfo.InvariantCulture,
            $"fork ({Shape.Stages.Count} {(Shape.Stages.Count == 1 ? "stage" : "stages")})");
}
